### PR TITLE
VM NICS: Add ASG and LB input on VM Nic

### PR DESCRIPTION
**Pull request by MariusStorhaug**
# Change

Add ASG and LB input on VM Nic using the existing `ipConfiguration` property of the VM NIC.

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update (Wiki)

## Checklist

- [ ] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

File: arm/Microsoft.Compute/virtualMachines/.bicep/nested_networkInterface.bicep [hunk 1]
```diff
@@ -73,6 +73,8 @@ resource networkInterface 'Microsoft.Network/networkInterfaces@2021-03-01' = {
         subnet: {
           id: ipConfiguration.subnetId
         }
+        loadBalancerBackendAddressPools: contains(ipConfiguration, 'loadBalancerBackendAddressPools') ? ipConfiguration.loadBalancerBackendAddressPools : null
```
Comment (eriqua) [inline arm/Microsoft.Compute/virtualMachines/.bicep/nested_networkInterface.bicep:76]: I'd suggest to move the complexity of building the loadBalancerBackendAddressPools object here to simplify the usage and avoid asking for the "id" property in the parameter file, similarly to what done for the subnet above
```suggestion
        loadBalancerBackendAddressPools: contains(ipConfiguration, 'loadBalancerBackendAddressPools') ? { 
        "id": ipConfiguration.loadBalancerBackendAddressPools
        } : null
```

File: arm/Microsoft.Compute/virtualMachines/.bicep/nested_networkInterface.bicep [hunk 2]
```diff
@@ -73,6 +73,8 @@ resource networkInterface 'Microsoft.Network/networkInterfaces@2021-03-01' = {
...
+        applicationSecurityGroups: contains(ipConfiguration, 'applicationSecurityGroups') ? ipConfiguration.applicationSecurityGroups : null
```
Comment (eriqua) [inline arm/Microsoft.Compute/virtualMachines/.bicep/nested_networkInterface.bicep:77]:
```suggestion
        applicationSecurityGroups: contains(ipConfiguration, 'applicationSecurityGroups') ? {
        "id": ipConfiguration.applicationSecurityGroups
        } : null
```

File: arm/Microsoft.Compute/virtualMachines/.parameters/linux.parameters.json [hunk 3]
```diff
@@ -73,7 +73,17 @@
                                         ]
                                     }
                                 ]
-                            }
+                            },
+                            "loadBalancerBackendAddressPools": [
+                                {
+                                    "id": "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Network/loadBalancers/adp-<<namePrefix>>-az-lb-x-001/backendAddressPools/servers"
```
Comment (eriqua) [inline arm/Microsoft.Compute/virtualMachines/.parameters/linux.parameters.json:79]: @
```suggestion
                                    "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Network/loadBalancers/adp-<<namePrefix>>-az-lb-x-001/backendAddressPools/servers"
```

File: arm/Microsoft.Compute/virtualMachines/.parameters/linux.parameters.json [hunk 4]
```diff
@@ -73,7 +73,17 @@
                                         ]
                                     }
                                 ]
-                            }
+                            },
+                            "loadBalancerBackendAddressPools": [
+                                {
```
Comment (eriqua) [inline arm/Microsoft.Compute/virtualMachines/.parameters/linux.parameters.json:78]:
```suggestion
```

File: arm/Microsoft.Compute/virtualMachines/.parameters/linux.parameters.json [hunk 5]
```diff
@@ -73,7 +73,17 @@
...
+                                    "id": "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Network/loadBalancers/adp-<<namePrefix>>-az-lb-x-001/backendAddressPools/servers"
+                                }
```
Comment (eriqua) [inline arm/Microsoft.Compute/virtualMachines/.parameters/linux.parameters.json:80]:
```suggestion
```

File: arm/Microsoft.Compute/virtualMachines/.parameters/linux.parameters.json [hunk 6]
```diff
@@ -73,7 +73,17 @@
...
+                            ],
+                            "applicationSecurityGroups": [
+                                {
+                                    "id": "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Network/applicationSecurityGroups/adp-<<namePrefix>>-az-asg-x-001"
```
Comment (eriqua) [inline arm/Microsoft.Compute/virtualMachines/.parameters/linux.parameters.json:84]:
```suggestion
"/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Network/applicationSecurityGroups/adp-<<namePrefix>>-az-asg-x-001"
```

File: arm/Microsoft.Compute/virtualMachines/.parameters/linux.parameters.json [hunk 7]
```diff
@@ -73,7 +73,17 @@
...
                                 ]
-                            }
+                            },
+                            "loadBalancerBackendAddressPools": [
+                                {
+                                    "id": "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Network/loadBalancers/adp-<<namePrefix>>-az-lb-x-001/backendAddressPools/servers"
+                                }
+                            ],
+                            "applicationSecurityGroups": [
+                                {
```
Comment (eriqua) [inline arm/Microsoft.Compute/virtualMachines/.parameters/linux.parameters.json:83]:
```suggestion
```

File: arm/Microsoft.Compute/virtualMachines/.parameters/linux.parameters.json [hunk 8]
```diff
@@ -73,7 +73,17 @@
...
+                                    "id": "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Network/applicationSecurityGroups/adp-<<namePrefix>>-az-asg-x-001"
+                                }
```
Comment (eriqua) [inline arm/Microsoft.Compute/virtualMachines/.parameters/linux.parameters.json:85]:
```suggestion
```

File: arm/Microsoft.Compute/virtualMachines/.parameters/windows.parameters.json [hunk 9]
```diff
@@ -62,7 +62,18 @@
                                         ]
                                     }
                                 ]
-                            }
+                            },
+                            "loadBalancerBackendAddressPools": [
+                                {
```
Comment (eriqua) [inline arm/Microsoft.Compute/virtualMachines/.parameters/windows.parameters.json:67]: Same here

File: arm/Microsoft.Compute/virtualMachines/.bicep/nested_networkInterface.bicep (hunk 1 above)
Comment (MariusStorhaug) [reply arm/Microsoft.Compute/virtualMachines/.bicep/nested_networkInterface.bicep:76]: @eriqua : What do we do here? Ref our discussion, its a var with nested loops if we want to do this. And that might be difficult to achieve. Thoughts?

Comment (MrMCake) [reply arm/Microsoft.Compute/virtualMachines/.bicep/nested_networkInterface.bicep:76]: Agreed. If it wouldn't be nested I'd agree

**Review by MrMCake — APPROVED** [review]

**Pull request merged**